utils/full_math.go

```diff
@@ -30,6 +30,127 @@ func MulDivRoundingUp(a, b, denominator *uint256.Int) (*uint256.Int, error) {
 	return resultU, nil
 }
 
+func MulDivRoundingUpV2(a, b, denominator, result *uint256.Int) error {
+	var remainder Uint256
+	err := MulDivV2(a, b, denominator, result, &remainder)
+	if err != nil {
+		return err
+	}
+
+	if !remainder.IsZero() {
+		if result.Cmp(MaxUint256) == 0 {
+			return ErrInvariant
+		}
+		result.AddUint64(result, 1)
+	}
+	return nil
+}
+
+// result=floor(a×b÷denominator), remainder=a×b%denominator
+// (pass remainder=nil if not required)
+func MulDivV2(a, b, denominator, result, remainder *uint256.Int) error {
```

Comment:
It is kind of difficult for me to realize that `remainder` is an optimization to skip and re-calculate `mulmod(a, b, denominator)` from the caller, especially when the `mulDiv` function in UniswapV3 contract doesn't return this value.

I would really appreciate it if we have a few comments to explain this optimization.

Reply:
Really sorry for that! I'll add comment here and update part 1 PR description as well, please let me know if there is anything unclear there, thanks!